### ISSUE TITLE
[QUESTION] Load more

### ISSUE THREAD

**Issue by angelru** [Keflon/FunctionZero.Maui.Controls#3]
Is there any thought to add the functionality of loading more elements?

**Comment by Keflon** [comment]
I have a bit of time on my hands next week so I can take a look.

**Comment by angelru** [comment]
Thanks!

**Comment by vhugogarcia** [comment]
I would love also this feature! Please.

**Comment by Keflon** [comment]
@angelru @vhugogarcia 
I've had a go at copying what I *think* the `CollectionView` `RemainingItemsThreshold` and `RemainingItemsThresholdReachedCommand` do, but it's problematic, for example:
- If there aren't enough items to fill the initial view then no callback is raised. 
- Issues with recursive callbacks if items are added asynchronously whilst scrolling, especially one-at-a-time with a delay between each.   

I'll take a look at how the `CollectionView` gets over these problems, and do something similar. If (as I suspect) the `CollectionView` also has these problems, then I'll think of a better way.  

Currently I'm thinking of just providing `SomethingHappenedCommand` and `SomethingHappenedEvent`. I'll find a better name when I've written it.   

These will fire every time something comes into view or goes out of view, including when items are initially rendered. It'll provide a payload that describes what's going on, something like:
```csharp
int NumItemsInView;
int FirstVisibleIndex;
int LastVisibleIndex;
...
int RemainingItems;
```
This will allow anyone to manage the ViewModel collection as they see fit, and keeps the door open to some sort of culling of items that are way out of view, as well as 'LoadMore'  when scrolling up.  

In this example, you could load more items if `RemainingItems` gets below some threshold. If you load items asynchronously you'd have to guard against recusive callbacks when scrolling, whilst ensuring you request only enough new items to keep your read-ahead threshold satisfied *after counting any pending asynchronous data*.

If you think the `SomethingHappenedCommand / Event` might do what you need, let me know and I can publish a preview package in a day or so for you to try.

**Comment by angelru** [comment]
> @angelru @vhugogarcia He intentado copiar lo que pienso y _hago_`CollectionView` `RemainingItemsThreshold` , `RemainingItemsThresholdReachedCommand`pero es problemático, por ejemplo:
> 
> * Si no hay suficientes elementos para llenar la vista inicial, no se genera ninguna devolución de llamada.
> * Problemas con las devoluciones de llamada recursivas si los elementos se agregan de forma asíncrona mientras se desplaza, especialmente uno a la vez con un retraso entre cada uno.
> 
> Echaré un vistazo a cómo `CollectionView`supera estos problemas y haré algo similar. Si (como sospecho) `CollectionView`también tiene estos problemas, pensaré en una mejor manera.
> 
> Actualmente estoy pensando en solo proporcionar `SomethingHappenedCommand`y `SomethingHappenedEvent`. Encontraré un nombre mejor cuando lo haya escrito.
> 
> Estos se activarán cada vez que algo aparezca o desaparezca, incluso cuando los elementos se rendericen inicialmente. Proporcionará una carga útil que describe lo que está pasando, algo como:
> 
> ```cs
> int NumItemsInView;
> int FirstVisibleIndex;
> int LastVisibleIndex;
> ...
> int RemainingItems;
> ```
> 
> Esto permitirá que cualquier persona administre la colección ViewModel como mejor le parezca, y mantiene la puerta abierta a algún tipo de selección de elementos que están fuera de la vista, así como 'Cargar más' al desplazarse hacia arriba.
> 
> En este ejemplo, podría cargar más elementos si `RemainingItems`se encuentra por debajo de algún umbral. Si carga elementos de forma asíncrona, debe protegerse contra las devoluciones de llamada recursivas al desplazarse, mientras se asegura de solicitar solo suficientes elementos nuevos para mantener satisfecho su umbral de lectura anticipada _después de contar los datos asíncronos pendientes_ .
> 
> Si cree que `SomethingHappenedCommand / Event`podría hacer lo que necesita, hágamelo saber y puedo publicar un paquete de vista previa en un día más o menos para que lo pruebe.

I seem to remember that the `RemainingItemsThresholdReachedCommand `was initialized even if the `RemainingItemsThreshold `was not met, I think the first time I solved it with some flag.

About:
Issues with recursive callbacks if items are added asynchronously while scrolling, especially one-at-a-time with a delay between each.

How about adding some `SemaphoreSlim `for that callback asynchronously? Although I don't know if that should be in the asynchronous call that executes SomethingHappenedCommand.

Avoid multiple executions of SomethingHappenedCommand? was what you mean?

**Comment by vhugogarcia** [comment]
What @angelru mentioned is totally right, in the CollectionView there is an issue when you reach the threshold, the UI list freezes waiting for the other elements to load, so they can be rendered. This shows a very bad experience for the users.

I believe this also must have somehow <LoadingIndicatorTemplate></LoadingIndicatorTemplate> where this template is being shown while the list is being appended with more items. Just a loud thinking.

Take as example: The Youtube mobile app on iOS or Android, see how smooth the scrolling is when loading more items to the list. I have no idea behind the scenes how they achieve that, but you cannot even see a loading indicator or UI freezes. It is just smooth and nice!

I believe this FunctionZero Maui control from @Keflon can achieve that. If this control is capable to load a list like the youtube with text and **images** fast and keep loading and loading fast then this will be the fastest list view on MAUI in the world.

**Comment by Keflon** [comment]
I've added `RemainingItems` and `RemainingItemsChangedCommand` to the `ListViewZero`.  This is in the GIT repo only, I hope to update the NuGet package next weekend.  
Take a look at [LazyListPageVm](https://github.com/Keflon/FunctionZero.Maui.Controls/blob/master/SampleApp/Mvvm/PageViewModels/List/LazyListPageVm.cs) for an example of how it works.  The corresponding UI is in [LazyListPage.xaml.](https://github.com/Keflon/FunctionZero.Maui.Controls/blob/master/SampleApp/Mvvm/Pages/List/LazyListPage.xaml)
Note that if there are not enough items to fill the view you'll get a negative number back, so for example '-3' means you need another 3 items to fill the view.  
The example aims to keep a minimum of 10 'remaining items' but adds new items one at a time and does so slowly, so we can see how to handle incoming changes to `RemainingItems` whilst some background process may be acquiring new items at its own pace.  

Is this what you had in mind? I hope it is useful.  

P.S. Note that WinUI has layout issues that prevent the ListViewZero updating the scroll-bar when new items are added, unless you resize the window. That's a WinUI bug, and there are multiple reports in the MAUI bug backlog.

**Comment by Keflon** [comment]
I've published [2.3.4-pre1](https://www.nuget.org/packages/FunctionZero.Maui.Controls/2.3.4-pre1) with the above changes as well as a couple of new controls, so you can try them out if you prefer sourcing from NuGet.

**Comment by angelru** [comment]
I like how it works, thank you very much!
